File: src/dna_predict/baseline.py
"""The dinucleotide baseline model.

EPIC's entry bar is beating a *dinucleotide* baseline: a model whose only
knowledge of the sequence is short-range base composition. Clearing it (plus a
method write-up) is what qualifies a team for consortium authorship, so we want
a faithful, honest implementation of it to measure ourselves against.

Design
------
For each strand we build a predictor keyed on the k-mer of length `k` ending at
(or centered on) a position. For k=2 this is the dinucleotide context. During
`fit` we accumulate, per k-mer, the mean observed initiation signal at all
positions sharing that context on the training contigs. At `predict` time we
emit, for every position, the learned mean for its local k-mer.

This captures exactly what a composition-only model can: some contexts (e.g.
around canonical initiator / TATA-like motifs) carry more initiation on
average than others. It has no notion of position, distance, or long-range
grammar -- which is the point.

The default k=2 is the literal dinucleotide baseline. `k` is exposed so we can
sanity-check how much pure short-range composition buys (k=1..4) without
turning this into a "real" model.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict

import numpy as np

from .data import InitiationTrack, encode_bases

Strand = str

# reverse-complement base index mapping: A<->T (0<->3), C<->G (1<->2)
_RC_IDX = np.array([3, 2, 1, 0], dtype=np.int8)


def _kmer_ids(base_idx: np.ndarray, k: int) -> np.ndarray:
    """Map each position to the integer id of the k-mer ending there.

    base_idx: int8 array of base indices in {0,1,2,3}, -1 for non-ACGT.
    Returns an int32 array of the same length; positions whose k-mer window
    contains a non-ACGT base or falls off the start get id -1 (unknown).
    """
    n = base_idx.shape[0]
    ids = np.full(n, -1, dtype=np.int32)
    if n < k:
        return ids
    # sliding window over the last k bases
    valid = np.ones(n, dtype=bool)
    acc = np.zeros(n, dtype=np.int64)
    for offset in range(k):
        shifted = np.full(n, -1, dtype=np.int64)
        if offset == 0:
            shifted[:] = base_idx
        else:
            shifted[offset:] = base_idx[:-offset]
        valid &= shifted >= 0
        acc = acc * 4 + np.where(shifted >= 0, shifted, 0)
    valid[: k - 1] = False  # no full window at the very start
    ids[valid] = acc[valid].astype(np.int32)
    return ids


@dataclass
class DinucleotideBaseline:
    """Composition-only initiation predictor (default: dinucleotide, k=2)."""

    k: int = 2
    # learned mean signal per k-mer id, per strand
    table_plus: np.ndarray = field(default_factory=lambda: np.zeros(0))
    table_minus: np.ndarray = field(default_factory=lambda: np.zeros(0))
    global_mean_plus: float = 0.0
    global_mean_minus: float = 0.0

    @property
    def n_kmers(self) -> int:
        return 4 ** self.k

    def _fit_one_strand(
        self, sequences: Dict[str, str], signal: Dict[str, np.ndarray]
    ) -> tuple[np.ndarray, float]:
        sums = np.zeros(self.n_kmers, dtype=np.float64)
        counts = np.zeros(self.n_kmers, dtype=np.float64)
        total = 0.0
        n_pos = 0
        for contig, seq in sequences.items():
            if contig not in signal:
                continue
            y = signal[contig]
            ids = _kmer_ids(encode_bases(seq), self.k)
            m = min(ids.shape[0], y.shape[0])
            ids, yv = ids[:m], y[:m]
            known = ids >= 0
            np.add.at(sums, ids[known], yv[known])
            np.add.at(counts, ids[known], 1.0)
            total += float(yv[known].sum())
            n_pos += int(known.sum())
        global_mean = total / n_pos if n_pos else 0.0
        # unseen k-mers fall back to the global mean
        table = np.divide(
            sums, counts, out=np.full(self.n_kmers, global_mean), where=counts > 0
        )
        return table.astype(np.float32), global_mean

    def fit(self, sequences: Dict[str, str], track: InitiationTrack) -> "DinucleotideBaseline":
        """Learn per-kmer mean initiation on the training contigs, per strand."""
        self.table_plus, self.global_mean_plus = self._fit_one_strand(sequences, track.plus)
        self.table_minus, self.global_mean_minus = self._fit_one_strand(sequences, track.minus)
        return self

    def predict_contig(self, seq: str, strand: Strand) -> np.ndarray:
        """Predict per-position initiation for one contig and strand."""
        if strand == "+":
            table, fallback = self.table_plus, self.global_mean_plus
        else:
            table, fallback = self.table_minus, self.global_mean_minus
        ids = _kmer_ids(encode_bases(seq), self.k)
        out = np.full(ids.shape[0], fallback, dtype=np.float32)
        known = ids >= 0
        out[known] = table[ids[known]]
        return out

    def predict(self, sequences: Dict[str, str]) -> InitiationTrack:
        """Predict a full strand-separated InitiationTrack for given contigs."""
        return InitiationTrack(
            plus={c: self.predict_contig(s, "+") for c, s in sequences.items()},
            minus={c: self.predict_contig(s, "-") for c, s in sequences.items()},
        )

File: rust/src/baseline.rs
//! The dinucleotide baseline model.
//!
//! EPIC's entry bar is beating a *dinucleotide* baseline: a model whose only
//! knowledge of the sequence is short-range base composition. Clearing it (plus
//! a method write-up) is what qualifies a team for consortium authorship, so we
//! want a faithful implementation to measure ourselves against.
//!
//! For each strand we learn, per k-mer of length `k` ending at a position, the
//! mean observed initiation signal across all training positions sharing that
//! context. At predict time we emit, for every position, the learned mean for
//! its local k-mer. `k = 2` is the literal dinucleotide baseline; `k` is exposed
//! only to sanity-check how much pure composition buys (k = 1..=4).

use std::collections::HashMap;

use crate::data::{encode_bases, InitiationTrack};
use crate::Strand;

/// Map each position to the integer id of the k-mer ending there.
///
/// `base_idx` holds base indices in 0..=3 with -1 for non-ACGT. Positions whose
/// k-mer window contains a non-ACGT base, or that fall within the first `k-1`
/// bases, get id -1 (unknown).
fn kmer_ids(base_idx: &[i8], k: usize) -> Vec<i32> {
    let n = base_idx.len();
    let mut ids = vec![-1i32; n];
    if n < k || k == 0 {
        return ids;
    }
    for i in (k - 1)..n {
        let mut acc: i32 = 0;
        let mut ok = true;
        for j in 0..k {
            let b = base_idx[i - (k - 1) + j];
            if b < 0 {
                ok = false;
                break;
            }
            acc = acc * 4 + b as i32;
        }
        if ok {
            ids[i] = acc;
        }
    }
    ids
}

/// Composition-only initiation predictor (default: dinucleotide, k = 2).
#[derive(Debug, Clone)]
pub struct DinucleotideBaseline {
    pub k: usize,
    table_plus: Vec<f32>,
    table_minus: Vec<f32>,
    global_mean_plus: f32,
    global_mean_minus: f32,
}

impl DinucleotideBaseline {
    pub fn new(k: usize) -> Self {
        let n_kmers = 4usize.pow(k as u32);
        Self {
            k,
            table_plus: vec![0.0; n_kmers],
            table_minus: vec![0.0; n_kmers],
            global_mean_plus: 0.0,
            global_mean_minus: 0.0,
        }
    }

    pub fn n_kmers(&self) -> usize {
        4usize.pow(self.k as u32)
    }

    fn fit_one_strand(
        &self,
        sequences: &HashMap<String, Vec<u8>>,
        signal: &HashMap<String, Vec<f32>>,
    ) -> (Vec<f32>, f32) {
        let n_kmers = self.n_kmers();
        let mut sums = vec![0.0f64; n_kmers];
        let mut counts = vec![0.0f64; n_kmers];
        let mut total = 0.0f64;
        let mut n_pos = 0usize;

        for (contig, seq) in sequences {
            let y = match signal.get(contig) {
                Some(y) => y,
                None => continue,
            };
            let ids = kmer_ids(&encode_bases(seq), self.k);
            let m = ids.len().min(y.len());
            for i in 0..m {
                let id = ids[i];
                if id >= 0 {
                    let id = id as usize;
                    let v = y[i] as f64;
                    sums[id] += v;
                    counts[id] += 1.0;
                    total += v;
                    n_pos += 1;
                }
            }
        }

        let global_mean = if n_pos > 0 {
            (total / n_pos as f64) as f32
        } else {
            0.0
        };
        // Unseen k-mers fall back to the global mean.
        let table: Vec<f32> = (0..n_kmers)
            .map(|i| {
                if counts[i] > 0.0 {
                    (sums[i] / counts[i]) as f32
                } else {
                    global_mean
                }
            })
            .collect();
        (table, global_mean)
    }

    /// Learn per-kmer mean initiation on the training contigs, per strand.
    pub fn fit(
        &mut self,
        sequences: &HashMap<String, Vec<u8>>,
        track: &InitiationTrack,
    ) -> &mut Self {
        let (tp, gp) = self.fit_one_strand(sequences, &track.plus);
        let (tm, gm) = self.fit_one_strand(sequences, &track.minus);
        self.table_plus = tp;
        self.global_mean_plus = gp;
        self.table_minus = tm;
        self.global_mean_minus = gm;
        self
    }

    /// Predict per-position initiation for one contig and strand.
    pub fn predict_contig(&self, seq: &[u8], strand: Strand) -> Vec<f32> {
        let (table, fallback) = match strand {
            Strand::Plus => (&self.table_plus, self.global_mean_plus),
            Strand::Minus => (&self.table_minus, self.global_mean_minus),
        };
        let ids = kmer_ids(&encode_bases(seq), self.k);
        ids.iter()
            .map(|&id| {
                if id >= 0 {
                    table[id as usize]
                } else {
                    fallback
                }
            })
            .collect()
    }

    /// Predict a full strand-separated track for the given contigs.
    pub fn predict(&self, sequences: &HashMap<String, Vec<u8>>) -> InitiationTrack {
        let mut out = InitiationTrack::new();
        for (contig, seq) in sequences {
            out.plus
                .insert(contig.clone(), self.predict_contig(seq, Strand::Plus));
            out.minus
                .insert(contig.clone(), self.predict_contig(seq, Strand::Minus));
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn kmer_ids_dinucleotide() {
        // sequence AC -> at pos1 the dinuc "AC" = 0*4 + 1 = 1
        let ids = kmer_ids(&[0, 1, 2, 3], 2);
        // pos0 has no full window -> -1; pos1 "AC"=1; pos2 "CG"=1*4+2=6; pos3 "GT"=2*4+3=11
        assert_eq!(ids, vec![-1, 1, 6, 11]);
    }

    #[test]
    fn unknown_base_yields_unknown_id() {
        let ids = kmer_ids(&[0, -1, 2], 2);
        assert_eq!(ids, vec![-1, -1, -1]);
    }
}
